FlowCrypt/Functionality/Mail Provider/Message Provider/MessageService.swift

```diff
@@ -38,19 +38,73 @@ extension FetchedMessage {
 }
 
 // MARK: - MessageService
+enum MessageServiceError: Error {
+    case missedPassPhrase(_ rawMimeData: Data)
+    case wrongPassPhrase(_ rawMimeData: Data, _ passPhrase: String)
+    // Could not fetch keys
+    case emptyKeys
+}
+
 final class MessageService {
     private let messageProvider: MessageProvider
-    private let dataService: DataServiceType & KeyDataServiceType
+    private let keyService: KeyServiceType
+    private let passPhraseStorage: PassPhraseStorageType
     private let core: Core
 
     init(
         messageProvider: MessageProvider = MailProvider.shared.messageProvider,
-        dataService: DataServiceType & KeyDataServiceType = DataService.shared,
-        core: Core = Core.shared
+        keyService: KeyServiceType = KeyService(),
+        core: Core = Core.shared,
+        passPhraseStorage: PassPhraseStorageType = PassPhraseStorage(
+            storage: EncryptedStorage(),
+            emailProvider: DataService.shared
+        )
     ) {
         self.messageProvider = messageProvider
-        self.dataService = dataService
+        self.keyService = keyService
         self.core = core
+        self.passPhraseStorage = passPhraseStorage
+    }
+
+    func validateMessage(rawMimeData: Data, with passPhrase: String) -> Promise<FetchedMessage> {
+        Promise<FetchedMessage> { [weak self] resolve, reject in
+            guard let self = self else { return }
+
+            guard let keys = try? self.keyService.getPrivateKeys(with: passPhrase).get(), keys.isNotEmpty else {
+                return reject(MessageServiceError.emptyKeys)
+            }
+
+            // TODO: - Tom - is it possible to get longid of the key which was used for decryption?
```

Comment:
@tomholub ?

Reply:
yes in principle. What is the ultimate goal?

Reply:
We pass array of keys `[key1-longid1, key2-longid2, key3-longid3]` to parse `parseDecryptMsg` in 
```
let decrypted = try self.core.parseDecryptMsg(
                encrypted: rawMimeData,
                keys: keys,
                msgPwd: nil,
                isEmail: true
)
```
If I get `longid` for the key which was used to decrypt a message I can save PassPhrase with proper longid here
```
keys
  .map { PassPhrase(value: passPhrase, longid: $0.longid) }
  .forEach { self.passPhraseStorage.savePassPhrase(with: $0, inStorage: false) }
```
Actually it's the same question as below - are there any possibility to check that entered pass phrase is matching particular key before I try to decrypt a message and check the error?

Reply:
let's continue discussion here https://github.com/FlowCrypt/flowcrypt-ios/issues/362